src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -198,820 +542,1177 @@ private void checkDataLength(int ... lengths) {
+                // number of bytes not filling a block
+                int remainder = bLen % blockSize;
+                // number of bytes along block boundary
+                bLen -= remainder;
+
+                // If there is enough bytes in ibuffer for a block or more,
+                // encrypt that first.
+                if (bLen > 0) {
+                    len += gctrghash.update(buffer, 0, bLen, out, outOfs);
+                    outOfs += bLen;
+                }
 
-        System.arraycopy(block, 0, out, (outOfs + len), tagLenBytes);
-        return (len + tagLenBytes);
-    }
+                // blen is now the offset for 'buffer'
+
+                // Construct and encrypt a block if there is enough 'buffer' and
+                // 'in' to make one
+                if ((inLen + remainder) >= blockSize) {
+                    byte[] block = new byte[blockSize];
+
+                    System.arraycopy(buffer, bLen, block, 0, remainder);
+                    int inLenUsed = blockSize - remainder;
+                    System.arraycopy(in, inOfs, block, remainder, inLenUsed);
+
+                    len += gctrghash.update(block, 0, blockSize, out, outOfs);
+                    inOfs += inLenUsed;
+                    inLen -= inLenUsed;
+                    outOfs += blockSize;
+                    ibuffer.reset();
+                    // Code below will write the remainder from 'in' to ibuffer
+                } else if (remainder > 0) {
+                    // If a block or more was encrypted from 'buffer' only, but
+                    // the rest of 'buffer' with 'in' could not construct a
+                    //  block, then put the rest of 'buffer' back into ibuffer.
+                    ibuffer.reset();
+                    ibuffer.write(buffer, bLen, remainder);
+                    // Code below will write the remainder from 'in' to ibuffer
+                }
+            }
 
-    int encryptFinal(ByteBuffer src, ByteBuffer dst)
-        throws IllegalBlockSizeException, ShortBufferException {
-        dst = overlapDetection(src, dst);
-        int len = src.remaining();
-        len += getBufferedLength();
+            // Encrypt the remaining blocks inside of 'in'
+            if (inLen > 0) {
+                len += gctrghash.update(in, inOfs, inLen, out, outOfs);
+            }
 
-        // 'len' includes ibuffer data
-        checkDataLength(len, tagLenBytes);
-        dst.mark();
-        if (dst.remaining() < len + tagLenBytes) {
-            throw new ShortBufferException("Output buffer too small");
-        }
+            // Write any remaining bytes less than a blockSize into ibuffer.
+            int remainder = inLen % blockSize;
+            if (remainder > 0) {
+                initBuffer(remainder);
+                inLen -= remainder;
+                // remainder offset is based on original buffer length
+                ibuffer.write(in, inOfs + inLen, remainder);
+            }
```

Comment:
Do you mean having all the GCM interface implementations have an argument that takes ibuffer and adds any unprocessed data into?  Maybe it would save a copy of the code, but I'm not sure I like GCTR or GHASH adding data into the ibuffer.

Reply:
Maybe there are other alternatives than making GCTR/GHASH to write into ibuffer. Or, perhaps we can figure out the right input length when passing them to GCTR or GHASH? This is again more for readability and robustness. What you have works, just need to be very careful and require detailed tracking/knowledge on GCTR/GHASH level since all inputs are passed down, but not all are used.